### ISSUE TITLE
Saving to parquet with partition_on results in loss of divisions when reading it back to dataframe.

### ISSUE THREAD

**Issue by TFelgueira** [dask/dask#3842]
# TLDR
When I save a Dask dataframe to parquet using partition on a certain collumn and then I read those parquet files into a dataframe the divisions are lost. This doesn't happen with every dataframe, for example if the collumn I'm using for the partitioning (let's call it 'Categoricals') is ordered then this problem won't happen. Below I'll leave an example where the problem happens and one where it doesn't.

# Imports

```python
import numpy as np
import pandas as pd
import dask.dataframe as dd
from fastparquet import ParquetFile
```

    /usr/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
      return f(*args, **kwds)
    /usr/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
      return f(*args, **kwds)

# Example where divisions are known

Here I created a dummy dataframe where the column by which we partition ('Categoricals') is ordered, as can be seen the divisions are known after loading the parquet file.

```python
def generate_df(size):
    df = pd.Series(np.arange(0, size), name='unique').to_frame()

    df['dates'] = pd.date_range('2015-01-01', periods=size, freq='1min', name='dates')
    df['ints'] = np.linspace(0, 100, size, endpoint=False, dtype=int)
    df['floats'] = df['ints'].astype(float)
    df['constants'] = 42
    df['uniques'] = df['unique'].astype(str)

    # nanints is ints with first 10% entries set to nan
    df['nanints'] = df['ints']
    df.loc[df.index < int(size / 10), 'nanints'] = np.nan

    # generate categorical column with 4 distinct values
    categoricals = []
    for value in ['a', 'b', 'c', 'd']:
        categoricals += [value] * int(size / 4)
    df['categoricals'] = categoricals
    df['nancategoricals'] = df['categoricals'].replace('d', np.nan)

    return df
```

```python
df = generate_df(16)
```

```python
df = df.set_index('dates')
df = dd.from_pandas(df, npartitions=3)
```

```python
df.head(16, npartitions=-1)
```

</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>unique</th>
      <th>ints</th>
      <th>floats</th>
      <th>constants</th>
      <th>uniques</th>
      <th>nanints</th>
      <th>categoricals</th>
      <th>nancategoricals</th>
    </tr>
    <tr>
      <th>dates</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2015-01-01 00:00:00</th>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>42</td>
      <td>0</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:01:00</th>
      <td>1</td>
      <td>6</td>
      <td>6.0</td>
      <td>42</td>
      <td>1</td>
      <td>6.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:02:00</th>
      <td>2</td>
      <td>12</td>
      <td>12.0</td>
      <td>42</td>
      <td>2</td>
      <td>12.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:03:00</th>
      <td>3</td>
      <td>18</td>
      <td>18.0</td>
      <td>42</td>
      <td>3</td>
      <td>18.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:04:00</th>
      <td>4</td>
      <td>25</td>
      <td>25.0</td>
      <td>42</td>
      <td>4</td>
      <td>25.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:05:00</th>
      <td>5</td>
      <td>31</td>
      <td>31.0</td>
      <td>42</td>
      <td>5</td>
      <td>31.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:06:00</th>
      <td>6</td>
      <td>37</td>
      <td>37.0</td>
      <td>42</td>
      <td>6</td>
      <td>37.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:07:00</th>
      <td>7</td>
      <td>43</td>
      <td>43.0</td>
      <td>42</td>
      <td>7</td>
      <td>43.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:08:00</th>
      <td>8</td>
      <td>50</td>
      <td>50.0</td>
      <td>42</td>
      <td>8</td>
      <td>50.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:09:00</th>
      <td>9</td>
      <td>56</td>
      <td>56.0</td>
      <td>42</td>
      <td>9</td>
      <td>56.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:10:00</th>
      <td>10</td>
      <td>62</td>
      <td>62.0</td>
      <td>42</td>
      <td>10</td>
      <td>62.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:11:00</th>
      <td>11</td>
      <td>68</td>
      <td>68.0</td>
      <td>42</td>
      <td>11</td>
      <td>68.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:12:00</th>
      <td>12</td>
      <td>75</td>
      <td>75.0</td>
      <td>42</td>
      <td>12</td>
      <td>75.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2015-01-01 00:13:00</th>
      <td>13</td>
      <td>81</td>
      <td>81.0</td>
      <td>42</td>
      <td>13</td>
      <td>81.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2015-01-01 00:14:00</th>
      <td>14</td>
      <td>87</td>
      <td>87.0</td>
      <td>42</td>
      <td>14</td>
      <td>87.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2015-01-01 00:15:00</th>
      <td>15</td>
      <td>93</td>
      <td>93.0</td>
      <td>42</td>
      <td>15</td>
      <td>93.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

```python
df.divisions 
```

    (Timestamp('2015-01-01 00:00:00'),
     Timestamp('2015-01-01 00:06:00'),
     Timestamp('2015-01-01 00:12:00'),
     Timestamp('2015-01-01 00:15:00'))

```python
df.to_parquet('./data_new', engine='fastparquet', partition_on=['categoricals'])

```

    /home/felgueira/.local/lib/python3.5/site-packages/fastparquet/writer.py:407: FutureWarning: Method .valid will be removed in a future version. Use .dropna instead.
      out = data.valid()  # better, data[data.notnull()], from above ?

```python
pq_f = ParquetFile('./data_new')

```

```python
df = dd.read_parquet(pq_f)

```

```python
df.divisions

```

    (Timestamp('2015-01-01 00:00:00'),
     Timestamp('2015-01-01 00:04:00'),
     Timestamp('2015-01-01 00:06:00'),
     Timestamp('2015-01-01 00:08:00'),
     Timestamp('2015-01-01 00:12:00'),
     Timestamp('2015-01-01 00:15:00'))

# Example where divisions are not known

Here I created a dummy dataframe where the column by which we partition ('Categoricals') is not ordered because of an "outlier", as can be seen the divisions are not known after loading the parquet file.

```python
def generate_df(size):
    df = pd.Series(np.arange(0, size), name='unique').to_frame()

    df['dates'] = pd.date_range('2015-01-01', periods=size, freq='1min', name='dates')
    df['ints'] = np.linspace(0, 100, size, endpoint=False, dtype=int)
    df['floats'] = df['ints'].astype(float)
    df['constants'] = 42
    df['uniques'] = df['unique'].astype(str)

    # nanints is ints with first 10% entries set to nan
    df['nanints'] = df['ints']
    df.loc[df.index < int(size / 10), 'nanints'] = np.nan

    # generate categorical column with 4 distinct values
    categoricals = []
    for value in ['a', 'b', 'c', 'd']:
        categoricals += [value] * int(size / 4)
    categoricals[10] = 'b'
    df['categoricals'] = categoricals
    df['nancategoricals'] = df['categoricals'].replace('d', np.nan)

    return df
```

```python
df = generate_df(16)
```

```python
df = df.set_index('dates')
df = dd.from_pandas(df, npartitions=3)
```

```python
df.head(16, npartitions=-1)
```

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>unique</th>
      <th>ints</th>
      <th>floats</th>
      <th>constants</th>
      <th>uniques</th>
      <th>nanints</th>
      <th>categoricals</th>
      <th>nancategoricals</th>
    </tr>
    <tr>
      <th>dates</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2015-01-01 00:00:00</th>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>42</td>
      <td>0</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:01:00</th>
      <td>1</td>
      <td>6</td>
      <td>6.0</td>
      <td>42</td>
      <td>1</td>
      <td>6.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:02:00</th>
      <td>2</td>
      <td>12</td>
      <td>12.0</td>
      <td>42</td>
      <td>2</td>
      <td>12.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:03:00</th>
      <td>3</td>
      <td>18</td>
      <td>18.0</td>
      <td>42</td>
      <td>3</td>
      <td>18.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:04:00</th>
      <td>4</td>
      <td>25</td>
      <td>25.0</td>
      <td>42</td>
      <td>4</td>
      <td>25.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:05:00</th>
      <td>5</td>
      <td>31</td>
      <td>31.0</td>
      <td>42</td>
      <td>5</td>
      <td>31.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:06:00</th>
      <td>6</td>
      <td>37</td>
      <td>37.0</td>
      <td>42</td>
      <td>6</td>
      <td>37.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:07:00</th>
      <td>7</td>
      <td>43</td>
      <td>43.0</td>
      <td>42</td>
      <td>7</td>
      <td>43.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:08:00</th>
      <td>8</td>
      <td>50</td>
      <td>50.0</td>
      <td>42</td>
      <td>8</td>
      <td>50.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:09:00</th>
      <td>9</td>
      <td>56</td>
      <td>56.0</td>
      <td>42</td>
      <td>9</td>
      <td>56.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:10:00</th>
      <td>10</td>
      <td>62</td>
      <td>62.0</td>
      <td>42</td>
      <td>10</td>
      <td>62.0</td>
      <td>b</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2015-01-01 00:11:00</th>
      <td>11</td>
      <td>68</td>
      <td>68.0</td>
      <td>42</td>
      <td>11</td>
      <td>68.0</td>
      <td>c</td>
      <td>c</td>
    </tr>
    <tr>
      <th>2015-01-01 00:12:00</th>
      <td>12</td>
      <td>75</td>
      <td>75.0</td>
      <td>42</td>
      <td>12</td>
      <td>75.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2015-01-01 00:13:00</th>
      <td>13</td>
      <td>81</td>
      <td>81.0</td>
      <td>42</td>
      <td>13</td>
      <td>81.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2015-01-01 00:14:00</th>
      <td>14</td>
      <td>87</td>
      <td>87.0</td>
      <td>42</td>
      <td>14</td>
      <td>87.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2015-01-01 00:15:00</th>
      <td>15</td>
      <td>93</td>
      <td>93.0</td>
      <td>42</td>
      <td>15</td>
      <td>93.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

```python
df.divisions 
```

    (Timestamp('2015-01-01 00:00:00'),
     Timestamp('2015-01-01 00:06:00'),
     Timestamp('2015-01-01 00:12:00'),
     Timestamp('2015-01-01 00:15:00'))

```python
df.to_parquet('./data_new', engine='fastparquet', partition_on=['categoricals'])

```

    /home/felgueira/.local/lib/python3.5/site-packages/fastparquet/writer.py:407: FutureWarning: Method .valid will be removed in a future version. Use .dropna instead.
      out = data.valid()  # better, data[data.notnull()], from above ?

```python
pq_f = ParquetFile('./data_new')

```

```python
df = dd.read_parquet(pq_f)

```

```python
df.divisions

```

    (None, None, None, None, None, None)

**Comment by birdsarah** [comment]
(Also this issue probably belongs in fastparquet repo?)

**Comment by jrbourbeau** [comment]
> Also this issue probably belongs in fastparquet repo?

@martindurant does the fastparquet repo seem like the appropriate place for this issue?

**Comment by martindurant** [comment]
This is not a fastparquet issue, no, but a fundamental limitation of partitioning in dask. I can imagine having better parquet-statistics-based optimisation down the line, but not immediately. This specific issue could be closed with docs clarification.
